db/blob_log_writer.cc

```diff
@@ -0,0 +1,219 @@
+{
+  assert (block_offset_ != 0);
+  assert (last_elem_type_ == ET_FILE_HDR || last_elem_type_ == ET_FOOTER);
+
+  std::string str;
+  footer.EncodeTo(&str);
+
+  Status s = dest_->Append(Slice(str));
+  if (s.ok())
+  {
+    block_offset_ += str.size();
+    s = dest_->Close();
+    dest_.reset();
+  }
+
+  last_elem_type_ = ET_FILE_FOOTER;
+  return s;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+//
+//
+////////////////////////////////////////////////////////////////////////////////
+Status Writer::AddRecord(const Slice& key, const Slice& val,
+  uint64_t* key_offset, uint64_t* blob_offset, uint32_t ttl)
+{
+  assert (block_offset_ != 0);
+  assert (last_elem_type_ == ET_FILE_HDR || last_elem_type_ == ET_FOOTER);
+
+  char buf[blob_log::BlobLogRecord::kHeaderSize];
+  ConstructBlobHeader(buf, key, val, ttl, -1);
+
+  Status s = EmitPhysicalRecord(buf, key, val, key_offset, blob_offset);
+  return s;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+//
+//
+////////////////////////////////////////////////////////////////////////////////
+Status Writer::AddRecord(const Slice& key, const Slice& val,
+  uint64_t* key_offset, uint64_t* blob_offset) {
+  assert (block_offset_ != 0);
+  assert (last_elem_type_ == ET_FILE_HDR || last_elem_type_ == ET_FOOTER);
+
+  char buf[blob_log::BlobLogRecord::kHeaderSize];
+  ConstructBlobHeader(buf, key, val, -1, -1);
+
+  Status s = EmitPhysicalRecord(buf, key, val, key_offset, blob_offset);
+  return s;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+//
+//
+////////////////////////////////////////////////////////////////////////////////
+void Writer::ConstructBlobHeader(char *headerbuf, const Slice& key,
+  const Slice& val, int32_t ttl, int64_t ts) {
+  RecordType t = kFullType;
+  RecordSubType st = kRegularType;
+  if (ttl != -1)
+    st = kTTLType;
+
+  uint32_t offset = 8;
+  // Format the header
+  EncodeFixed32(headerbuf+offset, key.size());
+  offset += 4;
+  EncodeFixed64(headerbuf+offset, val.size());
+  offset += 8;
+  if (ttl != -1) {
+    EncodeFixed32(headerbuf+offset, (uint32_t)ttl);
+  } else {
+    EncodeFixed32(headerbuf+offset, std::numeric_limits<uint32_t>::max());
+  }
+
+  offset += 4;
+  if (ts != -1) {
+    EncodeFixed64(headerbuf+offset, (uint64_t)ts);
+  } else {
+    EncodeFixed64(headerbuf+offset, std::numeric_limits<uint64_t>::max());
+  }
+  offset += 8;
+
+  headerbuf[offset] = static_cast<char>(t);
+  offset++;
+  headerbuf[offset] = static_cast<char>(st);
+
+  uint32_t header_crc = 0;
+  header_crc = crc32c::Extend(header_crc, headerbuf+2*sizeof(uint32_t),
+    blob_log::BlobLogRecord::kHeaderSize - 2*sizeof(uint32_t));
+  header_crc = crc32c::Mask(header_crc);
+  EncodeFixed32(headerbuf + 4, header_crc);
+
+  uint32_t crc = 0;
+  // Compute the crc of the record type and the payload.
+  crc = crc32c::Extend(crc, val.data(), val.size());
```

Comment:
added the key to the header_crc. want to keep blob separate because it is part of the Get calls.